File: src/pages/game/whoSaidWhat.nim
import std/[tables, strutils, strformat, options]
import ../../resources, ../../css/styles
import generator

const
    # Questions:
    idThesisDivPrefix: string = "id-thesis-div-"
    idThesisButtonsDiv: string = "id-thesis-buttons-"
    idThesisQuotePrefix: string = "id-thesis-quote-"
    idThesisAuthorPrefix: string = "id-thesis-author-"
    idThesisButtonAfdPrefix: string = "id-thesis-button-afd-"
    idThesisButtonOtherPrefix: string = "id-thesis-button-other-"

    # Score:
    idScoreAbsoluteLeft: string = "id-score-absolute-left"
    idScoreAbsoluteRight: string = "id-score-absolute-right"

    idScoreUniqueLeft: string = "id-score-unique-left"
    idScoreUniqueMiddle: string = "id-score-unique-middle"
    idScoreUniqueRight: string = "id-score-unique-right"

    # Buttons:
    idButtonStartQuestions: string = "id-button-start"
    idButtonSkipQuestion: string = "id-button-skip"
    idButtonNextQuestion: string = "id-button-next"


# Strings ---------------------------------------------------------------------

const strings = (
    meta: (
        title: lang("Who said what?", "Wer hat was gesagt?"),
        desc: lang(
            @[ # First element (split by "\n" used for meta description)
                "Who said what, an AfD member or another fascist? A depressing game about the (re-)rise of fascism in Germany.",
                "The source to the quote is featured after submitting your answer. If you discover an error or would like to submit more quotes, feel free to " & $a("https://github.com/nirokay/src.nirokay.com/blob/master/resources/game/who-said-what.json", "send a pull request on GitHub") & "."
            ].join("\n"),
            @[
                "Wer hat was gesagt, ein AfD Mitglied oder ein anderer Faschist? Ein deprimierendes Spiel über den (Wieder-)Anstieg des Faschismus in Deutschland.",
                "Die Quelle des Zitats is nach dem Einreichen einer Antwort zu Sehen. Falls Sie einen Fehler bemerken oder ein weiteres Zitat einreichen wollen, können Sie gerne " & $a("https://github.com/nirokay/src.nirokay.com/blob/master/resources/game/who-said-what.json", "ein Pull-Request via GitHub senden") & "."
            ].join("\n")
        )
    ),
    data: (
        source: lang("Source", "Quelle"),
        sourceMissing: lang("Source missing", "Quelle fehlt"),
        authorUnknown: lang("Unknown", "Unbekannt"),
        authorImgAlt: lang("Image of author", "Bild des Authors"),
        question: lang("Was this said by a member of the AfD?", "Wurde das von einem AfD-Mitglied gesagt?"),
        button: (
            yes: lang("Yes", "Ja"),
            no: lang("No", "Nein"),
            start: lang("Start", "Starten"),
            skip: lang("Skip", "Überspringen"),
            next: lang("Next", "Weiter")
        ),
        score: (
            correctUnique: lang("Score non-repeating questions: ", "Punkte nicht-wiederholter Fragen: "),
            correctAbsolute: lang("Score all questions: ", "Punkte aller Fragen: ")
        )
    )
)

# CSS -------------------------------------------------------------------------

var css: CssStyleSheet = newCssStyleSheet("game/who-said-what/who-said-what-styles.css")

const
    classQuestionBlock = ".thesis-question-div"{
        "border-color" := colourText,
        "background-color" := colourBackgroundMiddle,
        "border-radius" := "20px",
        "border-style" := "solid",
        "border-width" := "3px",
        "margin" := "20px",
        "padding" := "20px"
    }
    classWrongAnswer = ".thesis-wrong-answer"{
        "border-color" := colourRed
    }
    classCorrectAnswer = ".thesis-correct-answer"{
        "border-color" := colourGreen
    }
    classButton = ".true-false-button"{
        "color" := $Black
    }

    classBullshitQuote = ".quote-bullshit"{
        "font-size" := "1.5em"
    }

    classAuthorImage = ".author-image"{
        "max-width" := "200px",
        "max-height" := "200px"
    }

css.add(
    classQuestionBlock,
    classWrongAnswer,
    classCorrectAnswer,

    classButton,

    classBullshitQuote,

    classAuthorImage
)


# HTML ------------------------------------------------------------------------

var
    htmlEN: HtmlDocument
    htmlDE: HtmlDocument

proc getAuthor(thesis: WhoSaidWhatThesis): WhoSaidWhatAuthor =
    result = block:
        if whoSaidWhatJson.authors.hasKey(thesis.author):
            whoSaidWhatJson.authors[thesis.author]
        else:
            WhoSaidWhatAuthor()
proc getAuthorDiv(thesis: WhoSaidWhatThesis, id: int): HtmlElement =
    let author: WhoSaidWhatAuthor = thesis.getAuthor()
    let imgUrl: string = block:
        let path: string = author.imageUrl.get("unknown.svg")
        if path.startsWith("https://") or path.startsWith("http://") or path.startsWith("/"):
            path
        else:
            "/resources/images/games/who-said-what/" & path

    let allegiances: string = block:
        if author.allegiances.len() == 0:
            ""
        else:
            ": " & $i(author.allegiances.join(", "))

    let source: HtmlElement = p(
        if thesis.source.strip() == "":
            $strings.data.sourceMissing
        else:
            $a(thesis.source, $strings.data.source).addattr("target", "_blank")
    )

    result = `div`(
        img(imgUrl, $strings.data.authorImgAlt).setClass(classAuthorImage),
        p(thesis.author & allegiances),
        source
    ).setId(idThesisAuthorPrefix & $id).setStyle("display" := "none")

proc thesisHtmlBlock(id: int, thesis: WhoSaidWhatThesis): HtmlElement =
    let
        quote: LanguageString = lang(thesis.enGB, thesis.deDE)
        authorDiv: HtmlElement = thesis.getAuthorDiv(id)

        afdQuote: bool = block:
            var yes: bool = false
            for allegiance in thesis.getAuthor().allegiances:
                if allegiance.toLower() in ["afd", "ex-afd"]: yes = true
            yes

        buttonFunctionCorrect: string = &"submitQuestion({id}, {true});"
        buttonFunctionIncorrect: string = &"submitQuestion({id}, {false});"

        buttonWasSaidByAfd: string = if afdQuote: buttonFunctionCorrect else: buttonFunctionIncorrect
        buttonWasNotSaidByAfd: string = if not afdQuote: buttonFunctionCorrect else: buttonFunctionIncorrect

    result = `div`(
        q($quote).setId(idThesisQuotePrefix & $id).setClass(classBullshitQuote),
        `div`(
            authorDiv,
            `div`(
                p($strings.data.question),
                button($strings.data.button.yes, buttonWasSaidByAfd).setStyle("background-color" := colourGreen).setId(idThesisButtonAfdPrefix & $id).setClass(classButton),
                button($strings.data.button.no, buttonWasNotSaidByAfd).setStyle("background-color" := colourRed).setId(idThesisButtonOtherPrefix & $id).setClass(classButton)
            ).setId(idThesisButtonsDiv & $id)
        ).setClass(classFlexContainer).setStyle("align-items" := "center")
    ).setId(idThesisDivPrefix & $id).setClass(classQuestionBlock).setStyle("display" := "none")


for language in LANGUAGE:
    setTranslationTarget(language)
    var html: HtmlDocument = newHtmlLanguagedPage(
        $strings.meta.title,
        split($strings.meta.desc, "\n")[0],
        "who-said-what"
    )
    block `overwrite funky css file name stuff lol`:
        let cssOverride: CssStyleSheet = newCssStyleSheet(css.file.split("/")[^1])
        html.setStylesheet(cssOverride)
    html.addToHead(
        importScript("/javascript/game/who-said-what.js").addattr("defer")
    )

    html.add(
        header(
            h1($strings.meta.title),
            p($strings.meta.desc),
            `div`(
                p(
                    <$> $strings.data.score.correctAbsolute,
                    strong("0").setId(idScoreAbsoluteLeft),
                    <$> "/",
                    strong("0").setId(idScoreAbsoluteRight)
                ),
                p(
                    <$> $strings.data.score.correctUnique,
                    strong("0").setId(idScoreUniqueLeft),
                    <$> "/",
                    strong("0").setId(idScoreUniqueMiddle),
                    <$> "(",
                    <$> i("0").setId(idScoreUniqueRight),
                    <$> ")"
                )
            ).setClass(classFlexContainer),
            nav(
                button($strings.data.button.start, "whoSaidWhatStart();").setId(idButtonStartQuestions).setStyle("display" := "block"),
                button($strings.data.button.skip, "whoSaidWhatSkip();").setId(idButtonSkipQuestion).setStyle("display" := "none"),
                button($strings.data.button.next, "whoSaidWhatNext();").setId(idButtonNextQuestion).setStyle("display" := "none")
            ).setClass(classFlexContainer)
        )
    )

    var thesisList: seq[HtmlElement]
    for id, thesis in whoSaidWhatJson.thesis:
        thesisList.add thesisHtmlBlock(id, thesis)

    html.add(
        article(
            thesisList
        )
    )

    generateHtml()


incl htmlEN
incl htmlDE
incl css

File: src/javascript/game/who-said-what.ts
// Settings:
const questionsWithoutRepeating: number = 25;

// Questions:
const idThesisDivPrefix: string = "id-thesis-div-";
const idThesisButtonsDiv: string = "id-thesis-buttons-";
const idThesisQuotePrefix: string = "id-thesis-quote-";
const idThesisAuthorPrefix: string = "id-thesis-author-";
const idThesisButtonAfdPrefix: string = "id-thesis-button-afd-";
const idThesisButtonOtherPrefix: string = "id-thesis-button-other-";

// Score:
const idScoreAbsoluteLeft: string = "id-score-absolute-left";
const idScoreAbsoluteRight: string = "id-score-absolute-right";

const idScoreUniqueLeft: string = "id-score-unique-left";
const idScoreUniqueMiddle: string = "id-score-unique-middle";
const idScoreUniqueRight: string = "id-score-unique-right";

// Buttons:
const idButtonStartQuestions: string = "id-button-start";
const idButtonSkipQuestion: string = "id-button-skip";
const idButtonNextQuestion: string = "id-button-next";
const buttonIds: Array<string> = [idButtonStartQuestions, idButtonSkipQuestion, idButtonNextQuestion];

function displayOnlyButton(id: string) {
    buttonIds.forEach((id) => {
        let button: HTMLButtonElement|null = document.getElementById(id) as HTMLButtonElement;
        if(button != null && button != undefined) button.style.display = "none";
    });
    let button: HTMLButtonElement|null = document.getElementById(id) as HTMLButtonElement;
    if(button != null && button != undefined) button.style.display = "block";
}


let allQuestionsBuffer: Array<number> = [];
function isUnique(id: number): boolean {
    return ! allQuestionsBuffer.includes(id);
}
function addToQuestionBuffer(id: number) {
    if(isUnique(id)) allQuestionsBuffer.push(id);
}

function increaseScore(id: number, correct: boolean) {
    function readAndIncrement(element: HTMLElement|null) {
        if(element == null || element == undefined) {
            console.warn("Passed invalid element");
            return;
        }
        try {
            let rawOldValue: string = element.innerHTML;
            let newValue: number = parseInt(rawOldValue) + 1;
            element.innerHTML = newValue.toString();
        } catch(e) {
            console.error("Could not increment score for element", element, e);
            try {
                element.innerHTML = "0";
            } catch(e) {
                console.error("Could not reset score to 0", e);
            }
        }
    }
    let absoluteLeft: HTMLElement|null = document.getElementById(idScoreAbsoluteLeft);
    let absoluteRight: HTMLElement|null = document.getElementById(idScoreAbsoluteRight);
    if(correct) readAndIncrement(absoluteLeft);
    readAndIncrement(absoluteRight);

    let uniqueLeft: HTMLElement|null = document.getElementById(idScoreUniqueLeft);
    let uniqueMiddle: HTMLElement|null = document.getElementById(idScoreUniqueMiddle);
    let uniqueRight: HTMLElement|null = document.getElementById(idScoreUniqueRight);
    if(isUnique(id)) {
        if(correct) readAndIncrement(uniqueLeft);
        readAndIncrement(uniqueMiddle);
    }
    if(uniqueRight != null && uniqueRight != undefined) uniqueRight.innerHTML = everyGameQuestionId().length.toString();

    addToQuestionBuffer(id);
}


let lastQuestionsBuffer: Array<number> = [];
function addToLastQuestionBuffer(id: number) {
    lastQuestionsBuffer.push(id);
    if(lastQuestionsBuffer.length > questionsWithoutRepeating) lastQuestionsBuffer.shift();
}


function getQuestion(id: number): HTMLElement|null {
    return document.getElementById(idThesisDivPrefix + id.toString());
}
function getQuestionButtons(id: number): HTMLElement|null {
    return document.getElementById(idThesisButtonsDiv + id.toString());
}
function getQuestionAuthor(id: number): HTMLElement|null {
    return document.getElementById(idThesisAuthorPrefix + id.toString());
}


function everyGameQuestionId(): Array<number> {
    let result: Array<number> = [];
    for (let id = 0; id < 1024; id++) {
        let element: HTMLElement|null = getQuestion(id);
        if(element == null || element == undefined) break;
        result.push(id);
    }
    return result;
}


function hideAllQuestions() {
    everyGameQuestionId().forEach(id => {
        let question: HTMLElement|null = getQuestion(id);
        let author: HTMLElement|null = getQuestionAuthor(id);
        let buttons: HTMLElement|null = getQuestionButtons(id);
        if(question != null && question != undefined) question.style.display = "none";
        if(author != null && author != undefined) author.style.display = "none";
        if(buttons != null && buttons != undefined) buttons.style.display = "block";

        try {
            question?.classList.remove("thesis-wrong-answer");
            question?.classList.remove("thesis-correct-answer");
        } catch(e) {
            console.warn("Caught error whilst removing class", e);
        }
    });
}

function viewQuestionWithId(id: number) {
    hideAllQuestions();
    let element: HTMLElement|null = getQuestion(id);
    if(element != null && element != undefined) element.style.display = "block";
}
function viewRandomQuestion() {
    let ids: Array<number> = everyGameQuestionId();
    let randomIndex: number = -1;
    let tries: number = 0;
    while(lastQuestionsBuffer.includes(randomIndex) || randomIndex == -1) {
        // Give up after 1024 tries without recent picks:
        if(tries >= 1024) {
            alert("Please refresh the website, something seems to be broken... ran function 'viewRandomQuestion()' over 1024 times...");
            break;
        }
        tries++;
        let random: number = Math.random();
        if(random === 1) random = 0.99999;
        randomIndex = Math.floor(random * ids.length) % ids.length;
    }

    let id: number = ids[randomIndex];
    viewQuestionWithId(id);
    addToLastQuestionBuffer(id);
}

function submitQuestion(id: number, correct: boolean) {
    let question: HTMLElement|null = getQuestion(id);
    console.log(correct);
    if(question != null && question != undefined) {
        if(correct) {
            question.classList.remove("thesis-wrong-answer");
            question.classList.add("thesis-correct-answer");
        }
        else {
            question.classList.remove("thesis-correct-answer");
            question.classList.add("thesis-wrong-answer");
        }
    } else {
        alert("Fuck");
    }
    let buttons: HTMLElement|null = getQuestionButtons(id);
    if(buttons != null && buttons != undefined) buttons.style.display = "none";
    let author: HTMLElement|null = getQuestionAuthor(id);
    if(author != null && author != undefined) {
        author.style.display = "block";
    } else {
        console.warn("Could not display author", author);
    }

    increaseScore(id, correct);
    displayOnlyButton(idButtonNextQuestion);
}


function whoSaidWhatStart() {
    viewRandomQuestion();
    displayOnlyButton(idButtonSkipQuestion);

    let uniqueRight: HTMLElement|null = document.getElementById(idScoreUniqueRight);
    if(uniqueRight != null && uniqueRight != undefined) uniqueRight.innerHTML = everyGameQuestionId().length.toString();
}
function whoSaidWhatSkip() {
    viewRandomQuestion();
    displayOnlyButton(idButtonSkipQuestion);
}
function whoSaidWhatNext() {
    viewRandomQuestion();
    displayOnlyButton(idButtonSkipQuestion);
}


document.onload = () => {
    displayOnlyButton(idButtonStartQuestions);
    hideAllQuestions();
};
